

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>SVTK CLI - Command line tools for processing and analyzing structural variants &mdash; svtk 0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="svtk.utils - Utility function and classes for working with SV in Python" href="utils.html" />
    <link rel="prev" title="SVTK: A toolkit for manipulating and analyzing structural variants" href="index.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> svtk
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">SVTK CLI - Command line tools for processing and analyzing structural variants</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcfcluster">vcfcluster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#annotate">annotate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bedcluster">bedcluster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bincov">bincov</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#collect-pesr">collect-pesr</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#count-svtypes">count-svtypes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#resolve">resolve</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#standardize">standardize</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#vcf2bed">vcf2bed</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#remote-tabix">remote_tabix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#pe-test">pe-test</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sr-test">sr-test</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#count-pe">count-pe</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#count-sr">count-sr</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">svtk.utils - Utility function and classes for working with SV in Python</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">svtk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>SVTK CLI - Command line tools for processing and analyzing structural variants</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/cli.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="svtk-cli-command-line-tools-for-processing-and-analyzing-structural-variants">
<h1>SVTK CLI - Command line tools for processing and analyzing structural variants<a class="headerlink" href="#svtk-cli-command-line-tools-for-processing-and-analyzing-structural-variants" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="vcfcluster">
<h2>vcfcluster<a class="headerlink" href="#vcfcluster" title="Permalink to this headline">¶</a></h2>
<p><p>Intersect SV called by PE/SR-based algorithms.</p>
<p>Paired-end and split-read callers provide a reasonably precise estimation of
an SV breakpoint. This program identifies variant calls that fall within
the expected margin of error made by these programs and clusters them together.
The cluster distance defaults to 500 bp but it is recommended to use the
maximum individual clustering distance across the libraries being analyzed.
(Generally median + 7 * MAD)</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">vcfcluster</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span> <span class="n">REGION</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span> <span class="n">DIST</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span> <span class="n">FRAC</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">BED</span><span class="o">.</span><span class="n">GZ</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">-</span><span class="n">z</span> <span class="n">SVSIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="n">SVTYPES</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">SAMPLE_OVERLAP</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">preserve</span><span class="o">-</span><span class="n">ids</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">preserve</span><span class="o">-</span><span class="n">genotypes</span><span class="p">]</span>
                       <span class="n">filelist</span> <span class="n">fout</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>filelist</kbd></td>
<td>List of paths to standardized VCFS</td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td>Clustered VCF.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-r, --region</kbd></td>
<td>Restrict clustering to genomic region.</td></tr>
<tr><td class="option-group">
<kbd>-d, --dist</kbd></td>
<td><p class="first">Maximum clustering distance. Suggested to use max of median + 7*MAD over samples. [500]</p>
<p class="last">Default: 500</p>
</td></tr>
<tr><td class="option-group">
<kbd>-f, --frac</kbd></td>
<td><p class="first">Minimum reciprocal overlap between variants. [0.1]</p>
<p class="last">Default: 0.1</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-x, --blacklist</kbd></td>
</tr>
<tr><td>&#160;</td><td>Tabix indexed bed of blacklisted regions. Any SV with a breakpoint falling inside one of these regions is filtered from output.</td></tr>
<tr><td class="option-group">
<kbd>-z, --svsize</kbd></td>
<td><p class="first">Minimum SV size to report for intrachromosomal events. [0]</p>
<p class="last">Default: 500</p>
</td></tr>
<tr><td class="option-group">
<kbd>-p, --prefix</kbd></td>
<td><p class="first">Prefix for merged variant IDs. [MERGED]</p>
<p class="last">Default: “MERGED”</p>
</td></tr>
<tr><td class="option-group">
<kbd>-t, --svtypes</kbd></td>
<td><p class="first">Comma delimited list of svtypes to cluster [DEL,DUP,INV,BND]</p>
<p class="last">Default: “DEL,DUP,INV,BND”</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-o, --sample-overlap</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Minimum sample overlap for two variants to be clustered together.</p>
<p class="last">Default: 0.0</p>
</td></tr>
<tr><td class="option-group">
<kbd>--preserve-ids</kbd></td>
<td><p class="first">Include list of IDs of constituent records in each cluster.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--preserve-genotypes</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">In a set of clustered variants, report best (highest GQ) non-reference genotype when available.</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="annotate">
<h2>annotate<a class="headerlink" href="#annotate" title="Permalink to this headline">¶</a></h2>
<p><p>Annotate resolved SV with genic effects and noncoding hits.</p>
<p>This tool permits the user to provide a GTF of Gencode gene annotations, a BED
of noncoding elements, or both. The BED of noncoding elements must contain four
columns: (chrom, start, end, element_class).</p>
<p>The following classes of genic effects are annotated as new VCF INFO fields if
the SV meets the defined criteria:</p>
<blockquote>
<div><ol class="arabic">
<li><dl class="first docutils">
<dt>LOF - Loss of function.</dt>
<dd><ul class="first simple">
<li>Deletions are annotated LOF if they overlap any exon.</li>
<li>Duplications are annotated LOF if they reside entirely within</li>
</ul>
<p class="last">a gene boundary and overlap any exon.
* Inversions are annotated LOF if reside entirely within an exon, if
one breakpoint falls within an exon, if they reside entirely within a
gene boundary and overlap an exon, or if only one breakpoint falls
within a gene boundary.
* Translocations are annotated LOF If they fall within a gene boundary.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>COPY_GAIN</dt>
<dd><ul class="first simple">
<li>Duplications are annotated COPY_GAIN if they span the entirety of a</li>
</ul>
<p class="last">gene boundary.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>INTRONIC</dt>
<dd><ul class="first simple">
<li>Deletions, duplications, and inversions are annotated INTRONIC if</li>
</ul>
<p class="last">they are localized to an intron.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>DUP_PARTIAL</dt>
<dd><ul class="first simple">
<li>Duplications are annotated DUP_PARTIAL if they overlap the start or</li>
</ul>
<p class="last">end of a gene boundary but not its entirety, such that a whole copy of
the gene is preserved.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>INV_SPAN</dt>
<dd><ul class="first simple">
<li>Inversions are annotated INV_SPAN if they overlap the entirety of a</li>
</ul>
<p class="last">gene boundary without disrupting it.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>NEAREST_TSS</dt>
<dd><ul class="first simple">
<li>Intragenic events are annotated with the nearest transcription start</li>
</ul>
<p class="last">site.</p>
</dd>
</dl>
</li>
</ol>
</div></blockquote>
<p>An SV is annotated with a new NONCODING INFO field containing all classes of
noncoding elements which the variant overlaps.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">annotate</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gencode</span> <span class="n">GENCODE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">noncoding</span> <span class="n">NONCODING</span><span class="p">]</span>
                     <span class="n">vcf</span> <span class="n">annotated_vcf</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td>Structural variants.</td></tr>
<tr><td class="option-group">
<kbd>annotated_vcf</kbd></td>
<td>Annotated variants.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--gencode</kbd></td>
<td>Gencode gene annotations (GTF).</td></tr>
<tr><td class="option-group">
<kbd>--noncoding</kbd></td>
<td>Noncoding elements (bed). Columns = chr,start,end,element_class,element_name</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="bedcluster">
<h2>bedcluster<a class="headerlink" href="#bedcluster" title="Permalink to this headline">¶</a></h2>
<p><p>Cluster a bed of structural variants based on reciprocal overlap.</p>
<p>Variants are clustered with the single-linkage algorithm, using a minimum
reciprocal overlap to determine whether two variants are linked. If multiple
variants from the same sample are clustered together, they are merged into a
single spanning variant. Cluster compactness, calculated as the
root-mean-square standard deviation, is reported for each cluster. Optionally,
the median coordinates for each cluster are reported rather than the original
individual coordinates.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">bedcluster</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span> <span class="n">FRAC</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span> <span class="n">REGION</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">T</span> <span class="n">TMPDIR</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">INTERSECTION</span><span class="p">]</span>
                       <span class="n">bed</span> <span class="p">[</span><span class="n">fout</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>bed</kbd></td>
<td>SV calls to cluster. Columns: #chr, start, end, name, sample, svtype</td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td><p class="first">Clustered bed.</p>
<p class="last">Default: &lt;_io.TextIOWrapper name=’&lt;stdout&gt;’ mode=’w’ encoding=’UTF-8’&gt;</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-f, --frac</kbd></td>
<td><p class="first">Minimum reciprocal overlap fraction to link variants. [0.8]</p>
<p class="last">Default: 0.8</p>
</td></tr>
<tr><td class="option-group">
<kbd>-r, --region</kbd></td>
<td>Region to cluster (chrom:start-end). Requires tabixed bed.</td></tr>
<tr><td class="option-group">
<kbd>-p, --prefix</kbd></td>
<td><p class="first">Cluster ID prefix</p>
<p class="last">Default: “prefix”</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-m, --merge-coordinates</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Report median of start and end positions in each cluster as final coordinates of cluster.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>-T, --tmpdir</kbd></td>
<td>Temporary directory [/tmp]</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-s, --intersection</kbd></td>
</tr>
<tr><td>&#160;</td><td>Pre-computed self-intersection of bed.</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="bincov">
<h2>bincov<a class="headerlink" href="#bincov" title="Permalink to this headline">¶</a></h2>
<p>Calculates non-duplicate primary-aligned binned coverage
of a chromosome from an input BAM file</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">bincov</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">n</span> <span class="n">NORM_OUT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BINSIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span> <span class="p">{</span><span class="n">nucleotide</span><span class="p">,</span><span class="n">physical</span><span class="p">}]</span>
                   <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">BLACKLIST</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">z</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="n">OVERLAP</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">oldBT</span><span class="p">]</span>
                   <span class="n">bam</span> <span class="nb">chr</span> <span class="n">cov_out</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>bam</kbd></td>
<td>Input bam</td></tr>
<tr><td class="option-group">
<kbd>chr</kbd></td>
<td>Contig to evaluate</td></tr>
<tr><td class="option-group">
<kbd>cov_out</kbd></td>
<td>Output bed file of raw coverage</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-n, --norm_out</kbd></td>
<td>Output bed file of normalized coverage</td></tr>
<tr><td class="option-group">
<kbd>-b, --binsize</kbd></td>
<td><p class="first">Bin size (bp) [1000]</p>
<p class="last">Default: 1000</p>
</td></tr>
<tr><td class="option-group">
<kbd>-m, --mode</kbd></td>
<td><p class="first">Possible choices: nucleotide, physical</p>
<p>Type of coverage to calculate [nucleotide]</p>
<p class="last">Default: “nucleotide”</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-x, --blacklist</kbd></td>
</tr>
<tr><td>&#160;</td><td>BED file of regions to exclude</td></tr>
<tr><td class="option-group">
<kbd>-z, --gzip</kbd></td>
<td><p class="first">Compress output bed files</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-p, --presubsetted</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Input bam is already subsetted to desired chr</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>-v, --overlap</kbd></td>
<td><p class="first">Maximum fraction of each bin permitted to overlap with blacklist regions. [0.05]</p>
<p class="last">Default: 0.05</p>
</td></tr>
<tr><td class="option-group">
<kbd>--oldBT</kbd></td>
<td><p class="first">Using a bedtools version pre-2.24.0</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="collect-pesr">
<h2>collect-pesr<a class="headerlink" href="#collect-pesr" title="Permalink to this headline">¶</a></h2>
<p><p>Collect split read and discordant pair data from a bam alignment.</p>
<p>Split reads: The tool counts the number of reads soft-clipped in each direction
(30S121M = left-clipped, 121M30S = right-clipped) at each position in the
genome.  The position of a right-clipped read is shifted by the length of its
alignment.</p>
<p>Discordant pairs: The tool reduces discordant pairs to (chrA, posA, strandA,
chrB, posB, strandB).</p>
<p>Unmapped reads, reads with unmapped mates, secondary and supplementary
alignments, and duplicates are excluded (SAM flag 3340).</p>
<p>Collection can be performed on an S3-hosted bam. The tool will attempt to find
a local copy of the bam index in the working directory, or the directory
specified with <cite>–index-dir</cite>, otherwise the index will be downloaded.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">collect</span><span class="o">-</span><span class="n">pesr</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">index</span><span class="o">-</span><span class="nb">dir</span> <span class="n">INDEX_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span> <span class="n">REGION</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">z</span><span class="p">]</span>
                         <span class="n">bam</span> <span class="n">sample</span> <span class="n">splitfile</span> <span class="n">discfile</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>bam</kbd></td>
<td>Local or S3 path to bam</td></tr>
<tr><td class="option-group">
<kbd>sample</kbd></td>
<td>ID to append to each line of output files.</td></tr>
<tr><td class="option-group">
<kbd>splitfile</kbd></td>
<td>Output split counts.</td></tr>
<tr><td class="option-group">
<kbd>discfile</kbd></td>
<td>Output discordant pairs.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--index-dir</kbd></td>
<td>Directory of local BAM indexes if accessing a remote S3 bam.</td></tr>
<tr><td class="option-group">
<kbd>-r, --region</kbd></td>
<td>Tabix-formatted region to parse</td></tr>
<tr><td class="option-group">
<kbd>-z, --bgzip</kbd></td>
<td><p class="first">bgzip and tabix index output</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="count-svtypes">
<h2>count-svtypes<a class="headerlink" href="#count-svtypes" title="Permalink to this headline">¶</a></h2>
<p><p>Count the instances of each SVTYPE observed in each sample in a VCF.</p>
<p>A record is only counted towards a sample’s total if the sample received a
non-reference (0/0) or a non-null (./.) call. Records without an annotated
SVTYPE in the INFO field are counted as “NO_SVTYPE”.</p>
<p>Counts are reported in a three column table (sample, svtype, count).</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">count</span><span class="o">-</span><span class="n">svtypes</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">header</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">total</span><span class="o">-</span><span class="n">obs</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">total</span><span class="o">-</span><span class="n">variants</span><span class="p">]</span>
                          <span class="n">vcf</span> <span class="p">[</span><span class="n">fout</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td></td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td><p class="first">Output file [stdout]</p>
<p class="last">Default: &lt;_io.TextIOWrapper name=’&lt;stdout&gt;’ mode=’w’ encoding=’UTF-8’&gt;</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--no-header</kbd></td>
<td><p class="first">Don’t include header in output</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>--total-obs</kbd></td>
<td><p class="first">Sum variant counts across samples</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--total-variants</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Sum variant counts across samples</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="resolve">
<h2>resolve<a class="headerlink" href="#resolve" title="Permalink to this headline">¶</a></h2>
<p>Resolve complex SV from inversion/translocation breakpoints and CNV intervals.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">resolve</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">(</span><span class="o">--</span><span class="n">discfile</span> <span class="n">DISCFILE</span> <span class="o">|</span> <span class="o">--</span><span class="n">discfile</span><span class="o">-</span><span class="nb">list</span> <span class="n">DISCFILE_LIST</span><span class="p">)</span>
                    <span class="o">--</span><span class="n">mei</span><span class="o">-</span><span class="n">bed</span> <span class="n">MEI_BED</span> <span class="o">--</span><span class="n">cytobands</span> <span class="n">CYTOBANDS</span> <span class="p">[</span><span class="o">-</span><span class="n">u</span> <span class="n">UNRESOLVED</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">PREFIX</span><span class="p">]</span>
                    <span class="n">raw</span> <span class="n">resolved</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>raw</kbd></td>
<td>Filtered breakpoints and CNV intervals.</td></tr>
<tr><td class="option-group">
<kbd>resolved</kbd></td>
<td>Resolved simple and complex variants.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--discfile</kbd></td>
<td>Scraped discordant pairs. Required to attempt to resolve single-ender inversions.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--discfile-list</kbd></td>
</tr>
<tr><td>&#160;</td><td>Tab-delimited list of discordant pair files and indices</td></tr>
<tr><td class="option-group">
<kbd>--mei-bed</kbd></td>
<td>Mobile element insertion bed. Required to classify inverted insertions.</td></tr>
<tr><td class="option-group">
<kbd>--cytobands</kbd></td>
<td>Cytoband file. Required to correctly classify interchromosomal events.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-u, --unresolved</kbd></td>
</tr>
<tr><td>&#160;</td><td>Unresolved complex breakpoints and CNV.</td></tr>
<tr><td class="option-group">
<kbd>-p, --prefix</kbd></td>
<td><p class="first">Variant prefix [<a href="#id1"><span class="problematic" id="id2">CPX_</span></a>]</p>
<p class="last">Default: “<a href="#id3"><span class="problematic" id="id4">CPX_</span></a>”</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="standardize">
<h2>standardize<a class="headerlink" href="#standardize" title="Permalink to this headline">¶</a></h2>
<p><p>Standardize a VCF of SV calls.</p>
<p>Each record corresponds to a single SV breakpoint and will have the following
INFO fields, with specified constraints:</p>
<blockquote>
<div>SVTYPE:  SV type [DEL,DUP,INV,BND]
CHR2:    Secondary chromosome [Must be lexicographically greater than CHROM]
END:     SV end position (or position on CHR2 in translocations)
STRANDS: Breakpoint strandedness [++,+-,-+,–]
SVLEN:   SV length (-1 if translocation)
ALGORITHMS:  Source algorithm</div></blockquote>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">standardize</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">include</span><span class="o">-</span><span class="n">reference</span><span class="o">-</span><span class="n">sites</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">standardizer</span> <span class="n">STANDARDIZER</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">contigs</span> <span class="n">CONTIGS</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">size</span> <span class="n">MIN_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">call</span><span class="o">-</span><span class="n">null</span><span class="o">-</span><span class="n">sites</span><span class="p">]</span>
                        <span class="n">vcf</span> <span class="n">fout</span> <span class="n">source</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td>Raw VCF.</td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td>Standardized VCF.</td></tr>
<tr><td class="option-group">
<kbd>source</kbd></td>
<td>Source algorithm. [delly,lumpy,manta,wham,melt]</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-p, --prefix</kbd></td>
<td>If provided, variant names will be overwritten with this prefix.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--include-reference-sites</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Include records where all samples are called 0/0 or ./.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>--standardizer</kbd></td>
<td>Path to python file with custom standardizer definition. (Not yet supported.)</td></tr>
<tr><td class="option-group">
<kbd>--contigs</kbd></td>
<td>Reference fasta index (.fai). If provided, contigs in index will be used in VCF header. Otherwise all GRCh37 contigs will be used in header. Variants on contigs not in provided list will be removed.</td></tr>
<tr><td class="option-group">
<kbd>--min-size</kbd></td>
<td><p class="first">Minimum SV size to report [50].</p>
<p class="last">Default: 50</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--call-null-sites</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Call sites with null genotypes (./.). Generally useful when an algorithm has been run on a single sample and has only reported variant sites.</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="vcf2bed">
<h2>vcf2bed<a class="headerlink" href="#vcf2bed" title="Permalink to this headline">¶</a></h2>
<p>Convert a VCF to a BED.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">vcf2bed</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">samples</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">INFO</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">split</span><span class="o">-</span><span class="n">bnd</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">header</span><span class="p">]</span>
                    <span class="n">vcf</span> <span class="n">bed</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td>VCF to convert.</td></tr>
<tr><td class="option-group">
<kbd>bed</kbd></td>
<td>Converted bed. Specify <cite>-</cite> or <cite>stdout</cite> to write to stdout.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--no-samples</kbd></td>
<td><p class="first">Don’t include comma-delimited list of called samples for each variant.</p>
<p class="last">Default: True</p>
</td></tr>
<tr><td class="option-group">
<kbd>-i, --info</kbd></td>
<td>INFO field to include as column in output. May be specified more than once. To include all INFO fields, specify <cite>–info ALL</cite>. INFO fields are reported in the order in which they are requested. If ALL INFO fields are requested, they are reported in the order in which they appear in the VCF header.</td></tr>
<tr><td class="option-group">
<kbd>--split-bnd</kbd></td>
<td><p class="first">Report two entries in bed file for each BND.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>--no-header</kbd></td>
<td><p class="first">Suppress header.</p>
<p class="last">Default: True</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="remote-tabix">
<h2>remote_tabix<a class="headerlink" href="#remote-tabix" title="Permalink to this headline">¶</a></h2>
<p>Tabix into a remotely hosted file</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">remote</span><span class="o">-</span><span class="n">tabix</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">R</span> <span class="n">REGIONS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">header</span><span class="p">]</span> <span class="n">url</span> <span class="n">index</span> <span class="p">[</span><span class="n">region</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>url</kbd></td>
<td></td></tr>
<tr><td class="option-group">
<kbd>index</kbd></td>
<td></td></tr>
<tr><td class="option-group">
<kbd>region</kbd></td>
<td></td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-R, --regions</kbd></td>
<td>fetch all regions in bed file</td></tr>
<tr><td class="option-group">
<kbd>--header</kbd></td>
<td><p class="first">include header</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="pe-test">
<h2>pe-test<a class="headerlink" href="#pe-test" title="Permalink to this headline">¶</a></h2>
<p>Calculate enrichment of discordant pairs at SV breakpoints.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">pe</span><span class="o">-</span><span class="n">test</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">WINDOW_OUT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">WINDOW_IN</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BACKGROUND</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SAMPLES</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">index</span> <span class="n">INDEX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">medianfile</span> <span class="n">MEDIANFILE</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">log</span><span class="p">]</span>
                    <span class="n">vcf</span> <span class="n">disc</span> <span class="n">fout</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td>Variants.</td></tr>
<tr><td class="option-group">
<kbd>disc</kbd></td>
<td>Table of discordant pair coordinates.</td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td>Output table of PE counts.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd>-o, --window-out</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Window outside breakpoint to query for discordant pairs. [500]</p>
<p class="last">Default: 500</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-i, --window-in</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Window inside breakpoint to query for discordant pairs. [50]</p>
<p class="last">Default: 50</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-b, --background</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Number of background samples to sample for PE evidence. [160]</p>
<p class="last">Default: 160</p>
</td></tr>
<tr><td class="option-group">
<kbd>-s, --samples</kbd></td>
<td>Whitelist of samples to restrict testing to.</td></tr>
<tr><td class="option-group">
<kbd>--index</kbd></td>
<td>Tabix index of discordant pair file. Required if discordant pair file is hosted remotely.</td></tr>
<tr><td class="option-group">
<kbd>--medianfile</kbd></td>
<td>Median coverage statistics for each library (optional). If provided, each sample’s split counts will be normalized accordingly. Same format as RdTest, one column per sample.</td></tr>
<tr><td class="option-group">
<kbd>--log</kbd></td>
<td><p class="first">Print progress log to stderr.</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="sr-test">
<h2>sr-test<a class="headerlink" href="#sr-test" title="Permalink to this headline">¶</a></h2>
<p>Calculate enrichment of clipped reads at SV breakpoints.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">sr</span><span class="o">-</span><span class="n">test</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">w</span> <span class="n">WINDOW</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BACKGROUND</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SAMPLES</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">index</span> <span class="n">INDEX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">medianfile</span> <span class="n">MEDIANFILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">log</span><span class="p">]</span>
                    <span class="n">vcf</span> <span class="n">countfile</span> <span class="n">fout</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td>VCF of variant calls. Standardized to include CHR2, END, SVTYPE, STRANDS in INFO.</td></tr>
<tr><td class="option-group">
<kbd>countfile</kbd></td>
<td>Tabix indexed file of split counts. Columns: chrom,pos,clip,count,sample</td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td>Output table of most significant start/endpositions.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-w, --window</kbd></td>
<td><p class="first">Window around variant start/end to consider for split read support. [100]</p>
<p class="last">Default: 100</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-b, --background</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Number of background samples to choose for comparison in t-test. [160]</p>
<p class="last">Default: 160</p>
</td></tr>
<tr><td class="option-group">
<kbd>-s, --samples</kbd></td>
<td>Whitelist of samples to restrict testing to.</td></tr>
<tr><td class="option-group">
<kbd>--index</kbd></td>
<td>Tabix index of discordant pair file. Required if discordant pair file is hosted remotely.</td></tr>
<tr><td class="option-group">
<kbd>--medianfile</kbd></td>
<td>Median coverage statistics for each library (optional). If provided, each sample’s split counts will be normalized accordingly. Same format as RdTest, one column per sample.</td></tr>
<tr><td class="option-group">
<kbd>--log</kbd></td>
<td><p class="first">Print progress log to stderr.</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="count-pe">
<h2>count-pe<a class="headerlink" href="#count-pe" title="Permalink to this headline">¶</a></h2>
<p>Count discordant pairs supporting a SV breakpoints.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">count</span><span class="o">-</span><span class="n">pe</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">WINDOW_OUT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="n">WINDOW_IN</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SAMPLES</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">index</span> <span class="n">INDEX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">medianfile</span> <span class="n">MEDIANFILE</span><span class="p">]</span>
                     <span class="n">vcf</span> <span class="n">disc</span> <span class="n">fout</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td>Variants.</td></tr>
<tr><td class="option-group">
<kbd>disc</kbd></td>
<td>Table of discordant pair coordinates.</td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td>Output table of PE counts.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd>-o, --window-out</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Window outside breakpoint to query for discordant pairs. [500]</p>
<p class="last">Default: 500</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-i, --window-in</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Window inside breakpoint to query for discordant pairs. [50]</p>
<p class="last">Default: 50</p>
</td></tr>
<tr><td class="option-group">
<kbd>-s, --samples</kbd></td>
<td>Whitelist of samples to restrict testing to.</td></tr>
<tr><td class="option-group">
<kbd>--index</kbd></td>
<td>Tabix index of discordant pair file. Required if discordant pair file is hosted remotely.</td></tr>
<tr><td class="option-group">
<kbd>--medianfile</kbd></td>
<td>Median coverage statistics for each library (optional). If provided, each sample’s split counts will be normalized accordingly. Same format as RdTest, one column per sample.</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="count-sr">
<h2>count-sr<a class="headerlink" href="#count-sr" title="Permalink to this headline">¶</a></h2>
<p>Count clipped reads at SV breakpoints. Unwindowed.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">svtk</span> <span class="n">count</span><span class="o">-</span><span class="n">sr</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SAMPLES</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">index</span> <span class="n">INDEX</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">medianfile</span> <span class="n">MEDIANFILE</span><span class="p">]</span>
                     <span class="n">vcf</span> <span class="n">countfile</span> <span class="n">fout</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>vcf</kbd></td>
<td>VCF of variant calls. Standardized to include CHR2, END, SVTYPE, STRANDS in INFO.</td></tr>
<tr><td class="option-group">
<kbd>countfile</kbd></td>
<td>Tabix indexed file of split counts. Columns: chrom,pos,clip,count,sample</td></tr>
<tr><td class="option-group">
<kbd>fout</kbd></td>
<td>Output table of split read counts.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h3>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-s, --samples</kbd></td>
<td>Whitelist of samples to restrict testing to.</td></tr>
<tr><td class="option-group">
<kbd>--index</kbd></td>
<td>Tabix index of discordant pair file. Required if discordant pair file is hosted remotely.</td></tr>
<tr><td class="option-group">
<kbd>--medianfile</kbd></td>
<td>Median coverage statistics for each library (optional). If provided, each sample’s split counts will be normalized accordingly. Same format as RdTest, one column per sample.</td></tr>
</tbody>
</table>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="utils.html" class="btn btn-neutral float-right" title="svtk.utils - Utility function and classes for working with SV in Python" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="SVTK: A toolkit for manipulating and analyzing structural variants" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Matthew Stone.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.1',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>