

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>svtools package &mdash; svtools 0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="svtools 0.1 documentation" href="../../index.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> svtools
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">svtools package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-svtools.genomeslink">svtools.genomeslink module</a></li>
<li><a class="reference internal" href="#module-svtools.svfile">svtools.svfile module</a></li>
<li><a class="reference internal" href="#module-svtools.utils">svtools.utils module</a></li>
<li><a class="reference internal" href="#module-svtools">Module contents</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">svtools</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
      <li>svtools package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/docs/source/svtools.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="svtools-package">
<h1>svtools package<a class="headerlink" href="#svtools-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="svtools.standardize.html">svtools.standardize package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="svtools.standardize.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="svtools.standardize.html#module-svtools.standardize.standardize">svtools.standardize.standardize module</a></li>
<li class="toctree-l2"><a class="reference internal" href="svtools.standardize.html#module-svtools.standardize.std_delly">svtools.standardize.std_delly module</a></li>
<li class="toctree-l2"><a class="reference internal" href="svtools.standardize.html#module-svtools.standardize.std_lumpy">svtools.standardize.std_lumpy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="svtools.standardize.html#module-svtools.standardize.std_manta">svtools.standardize.std_manta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="svtools.standardize.html#module-svtools.standardize.std_wham">svtools.standardize.std_wham module</a></li>
<li class="toctree-l2"><a class="reference internal" href="svtools.standardize.html#module-svtools.standardize">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-svtools.genomeslink">
<span id="svtools-genomeslink-module"></span><h2>svtools.genomeslink module<a class="headerlink" href="#module-svtools.genomeslink" title="Permalink to this headline">¶</a></h2>
<p>genomeslink.py</p>
<p>Implementation of single-linkage clustering for genomic coordinates.</p>
<p>Not actually an implementation of the SLINK algorithm.</p>
<p>Copyright © 2015 Matthew Stone &lt;<a class="reference external" href="mailto:mstone5&#37;&#52;&#48;mgh&#46;harvard&#46;edu">mstone5<span>&#64;</span>mgh<span>&#46;</span>harvard<span>&#46;</span>edu</a>&gt;
Distributed under terms of the MIT license.</p>
<dl class="class">
<dt id="svtools.genomeslink.GSNode">
<em class="property">class </em><code class="descclassname">svtools.genomeslink.</code><code class="descname">GSNode</code><span class="sig-paren">(</span><em>chrA</em>, <em>posA</em>, <em>chrB</em>, <em>posB</em>, <em>name='.'</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GSNode" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="svtools.genomeslink.GSNode.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>chrA</em>, <em>posA</em>, <em>chrB</em>, <em>posB</em>, <em>name='.'</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GSNode.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Node in graph-based single-linkage clustering of genomic coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GSNode.clusters_with">
<code class="descname">clusters_with</code><span class="sig-paren">(</span><em>other</em>, <em>dist</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GSNode.clusters_with" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if candidates meet cluster distance requirement on chrB, posB</p>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GSNode.is_allowed_chrom">
<code class="descname">is_allowed_chrom</code><span class="sig-paren">(</span><em>chroms='either'</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GSNode.is_allowed_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Whitelist chromosomes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>chroms</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>optional</em>) &#8211; <ol class="arabic simple">
<li>List of permitted chromosomes</li>
</ol>
<p>2) One of &#8216;GRCh&#8217;, &#8216;UCSC&#8217;, &#8216;either&#8217;. Allows the 22 autosomes and sex
chromosomes in GRCh format (1, 2, ... X, Y), UCSC format
(chr1, chr2, ... chrX, chrY), or either format.
Defaults to &#8216;either&#8217;.</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>is_allowed_chrom</strong> &#8211; True if both reads are on whitelisted chromosomes.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GSNode.is_clusterable_with">
<code class="descname">is_clusterable_with</code><span class="sig-paren">(</span><em>other</em>, <em>dist</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GSNode.is_clusterable_with" title="Permalink to this definition">¶</a></dt>
<dd><p>Identify cluster candidates if chrA, posB meet cluster distance limit</p>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GSNode.is_in">
<code class="descname">is_in</code><span class="sig-paren">(</span><em>tabixfile</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GSNode.is_in" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if breakpoints of SV fall into any region in tabix-indexed bed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>tabixfile</strong> (<em>pysam.TabixFile</em>) &#8211; </td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>is_in</strong></td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="svtools.genomeslink.GSNode.secondary">
<code class="descname">secondary</code><a class="headerlink" href="#svtools.genomeslink.GSNode.secondary" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter function</p>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GSNode.sort_positions">
<code class="descname">sort_positions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GSNode.sort_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Force chrA, posA to be upstream of chrB, posB</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="svtools.genomeslink.GenomeSLINK">
<em class="property">class </em><code class="descclassname">svtools.genomeslink.</code><code class="descname">GenomeSLINK</code><span class="sig-paren">(</span><em>nodes</em>, <em>dist</em>, <em>size=1</em>, <em>blacklist=None</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GenomeSLINK" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="svtools.genomeslink.GenomeSLINK.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>nodes</em>, <em>dist</em>, <em>size=1</em>, <em>blacklist=None</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GenomeSLINK.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Graph-based single-linkage clustering of genomic coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>nodes</strong> (<em>iterable of GSNode</em>) &#8211; GSNodes sorted by chrA, posA</li>
<li><strong>dist</strong> (<em>int</em>) &#8211; Maximum clustering distance.</li>
<li><strong>size</strong> (<em>int</em>) &#8211; Minimum cluster size. Recommended to use 1 for call/variant
clustering, scale up for read pair clustering.</li>
<li><strong>blacklist</strong> (<em>pysam.TabixFile</em><em>, </em><em>optional</em>) &#8211; Regions to exclude from clustering. Any node with a coordinate
inside an excluding region is omitted. (NOTE: not overlap-based.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GenomeSLINK.cluster">
<code class="descname">cluster</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GenomeSLINK.cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform single linkage clustering on a candidate batch of GSNodes</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Yields:</th><td class="field-body"><em>list of GSNode</em> &#8211; A cluster of GSNodes</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GenomeSLINK.cluster_candidates">
<code class="descname">cluster_candidates</code><span class="sig-paren">(</span><em>candidates</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GenomeSLINK.cluster_candidates" title="Permalink to this definition">¶</a></dt>
<dd><p>Batch of clustering</p>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GenomeSLINK.filter_nodes">
<code class="descname">filter_nodes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GenomeSLINK.filter_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter provided nodes. By default, remove nodes in blacklisted regions.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Yields:</th><td class="field-body"><strong>node</strong> (<em>GSNode</em>)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="svtools.genomeslink.GenomeSLINK.get_candidates">
<code class="descname">get_candidates</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#svtools.genomeslink.GenomeSLINK.get_candidates" title="Permalink to this definition">¶</a></dt>
<dd><p>Find batches of SVCalls eligible for clustering.</p>
<p>Requires input sorted by chromosome and position of first read in each
pair. Pairs are collected while the first read in the next pair in the
parser is within the maximum clustering distance of the first read of
the previous pair.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Yields:</th><td class="field-body"><strong>candidates</strong> (<em>deque of GSNode</em>)</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-svtools.svfile">
<span id="svtools-svfile-module"></span><h2>svtools.svfile module<a class="headerlink" href="#module-svtools.svfile" title="Permalink to this headline">¶</a></h2>
<p>svfile.py</p>
<p>Wrap the pysam API to permit clustering of standardized SV VCF records.</p>
<p>Copyright © 2016 Matthew Stone &lt;<a class="reference external" href="mailto:mstone5&#37;&#52;&#48;mgh&#46;harvard&#46;edu">mstone5<span>&#64;</span>mgh<span>&#46;</span>harvard<span>&#46;</span>edu</a>&gt;
Distributed under terms of the MIT license.</p>
<dl class="class">
<dt id="svtools.svfile.SVFile">
<em class="property">class </em><code class="descclassname">svtools.svfile.</code><code class="descname">SVFile</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="svtools.svfile.SVFile.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVFile.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper for standardized VCF files.</p>
</dd></dl>

<dl class="method">
<dt id="svtools.svfile.SVFile.fetch">
<code class="descname">fetch</code><span class="sig-paren">(</span><em>chrom</em>, <em>start=None</em>, <em>end=None</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVFile.fetch" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch only calls from specified region of SVFile.</p>
<p>Requires Tabix index. Updates SVFile in place.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>chrom</strong> (<em>str</em>) &#8211; </li>
<li><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) &#8211; </li>
<li><strong>end</strong> (<em>int</em><em>, </em><em>optional</em>) &#8211; Required if start specified</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="svtools.svfile.SVFile.next">
<code class="descname">next</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVFile.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="svtools.svfile.SVRecord">
<em class="property">class </em><code class="descclassname">svtools.svfile.</code><code class="descname">SVRecord</code><span class="sig-paren">(</span><em>record</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVRecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#svtools.genomeslink.GSNode" title="svtools.genomeslink.GSNode"><code class="xref py py-class docutils literal"><span class="pre">svtools.genomeslink.GSNode</span></code></a></p>
<p>Clusterable VCF record.</p>
<dl class="method">
<dt id="svtools.svfile.SVRecord.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>record</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVRecord.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>record <span class="classifier-delimiter">:</span> <span class="classifier">pysam.VariantRecord</span></dt>
<dd>Must specify &#8216;CHR2&#8217; and &#8216;END&#8217; in INFO</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="svtools.svfile.SVRecord.clusters_with">
<code class="descname">clusters_with</code><span class="sig-paren">(</span><em>other</em>, <em>dist</em>, <em>frac=0.0</em>, <em>match_strands=False</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVRecord.clusters_with" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="svtools.svfile.SVRecord.is_tloc">
<code class="descname">is_tloc</code><a class="headerlink" href="#svtools.svfile.SVRecord.is_tloc" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="svtools.svfile.SVRecord.svtype">
<code class="descname">svtype</code><a class="headerlink" href="#svtools.svfile.SVRecord.svtype" title="Permalink to this definition">¶</a></dt>
<dd><p>returns: <strong>svtype</strong> &#8211; One of {DEL, DUP, INV, BND}
:rtype: str</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="svtools.svfile.SVRecordCluster">
<em class="property">class </em><code class="descclassname">svtools.svfile.</code><code class="descname">SVRecordCluster</code><span class="sig-paren">(</span><em>records</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVRecordCluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="svtools.svfile.SVRecordCluster.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>records</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVRecordCluster.__init__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="svtools.svfile.SVRecordCluster.merge">
<code class="descname">merge</code><span class="sig-paren">(</span><em>new_record</em>, <em>sources</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVRecordCluster.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Aggregate clustered records.</p>
<ul class="simple">
<li>Original record IDs are preserved in a new INFO field</li>
</ul>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>new_record</strong> (<em>pysam.VariantRecord</em>) &#8211; Blank record to fill with aggregated data</li>
<li><strong>sources</strong> (<em>list of str</em>) &#8211; List of all sources</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>new_record</strong> &#8211; VCF record populated with aggregate cluster data</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pysam.VariantRecord</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="svtools.svfile.SVRecordCluster.merge_pos">
<code class="descname">merge_pos</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#svtools.svfile.SVRecordCluster.merge_pos" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute aggregate POS/END of clustered SVRecords.</p>
<p>Defaults to computing median of POS and END over constituent
records. CIPOS/CIEND are calculated as (MIN - MEDIAN, MAX - MEDIAN)
over POS and END.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><ul class="simple">
<li><strong>POS</strong> (<em>int</em>)</li>
<li><strong>END</strong> (<em>int</em>)</li>
<li><strong>CIPOS</strong> (<em>list of int</em>)</li>
<li><strong>CIEND</strong> (<em>list of int</em>)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="svtools.svfile.SVRecordCluster.rmsstd">
<code class="descname">rmsstd</code><a class="headerlink" href="#svtools.svfile.SVRecordCluster.rmsstd" title="Permalink to this definition">¶</a></dt>
<dd><p>Root-mean-square standard deviation of cluster coordinates</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-svtools.utils">
<span id="svtools-utils-module"></span><h2>svtools.utils module<a class="headerlink" href="#module-svtools.utils" title="Permalink to this headline">¶</a></h2>
<p>utils.py</p>
<p>Helper functions for svtools.</p>
<p>Copyright © 2017 Matthew Stone &lt;<a class="reference external" href="mailto:mstone5&#37;&#52;&#48;mgh&#46;harvard&#46;edu">mstone5<span>&#64;</span>mgh<span>&#46;</span>harvard<span>&#46;</span>edu</a>&gt;
Distributed under terms of the MIT license.</p>
<dl class="function">
<dt id="svtools.utils.is_smaller_chrom">
<code class="descclassname">svtools.utils.</code><code class="descname">is_smaller_chrom</code><span class="sig-paren">(</span><em>chrA</em>, <em>chrB</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.utils.is_smaller_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if chrA is naturally less than chrB</p>
<p>Returns True if chrA == chrB so comparison will default to position</p>
</dd></dl>

<dl class="function">
<dt id="svtools.utils.make_bnd_alt">
<code class="descclassname">svtools.utils.</code><code class="descname">make_bnd_alt</code><span class="sig-paren">(</span><em>chrom</em>, <em>pos</em>, <em>strands</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.utils.make_bnd_alt" title="Permalink to this definition">¶</a></dt>
<dd><p>Make ALT for BND record in accordance with VCF specification.</p>
</dd></dl>

<dl class="function">
<dt id="svtools.utils.recip">
<code class="descclassname">svtools.utils.</code><code class="descname">recip</code><span class="sig-paren">(</span><em>startA</em>, <em>endA</em>, <em>startB</em>, <em>endB</em>, <em>frac</em><span class="sig-paren">)</span><a class="headerlink" href="#svtools.utils.recip" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if two intervals share a specified reciprocal overlap.</p>
</dd></dl>

</div>
<div class="section" id="module-svtools">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-svtools" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Matthew Stone.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>